### PR TITLE
pt: refact training code

### PR DESCRIPTION
**Pull request by iProzd**
This PR
- add data_requirement for dataloader 
- reformat `make_stat_input` and related training code
- support single-task & multi-task training

### CODE REVIEW

**Review by CaRoLZhangxy — APPROVED** [review]

File: deepmd/dpmodel/descriptor/make_base_descriptor.py [hunk 1]
```diff
@@ -84,8 +86,13 @@ def mixed_types(self) -> bool:
...
             pass
 
+        @abstractmethod
+        def share_params(self, base_class, shared_level, resume=False):
+            """Share the parameters of self to the base_class with shared_level."""
+            pass
+
         def compute_input_stats(
-            self, merged: List[dict], path: Optional[DPPath] = None
+            self, merged: Union[Callable, List[dict]], path: Optional[DPPath] = None
```
Comment (njzjz) [inline deepmd/dpmodel/descriptor/make_base_descriptor.py:95]: What is the purpose of this change?

Comment (njzjz) [reply deepmd/dpmodel/descriptor/make_base_descriptor.py:95]: Anyway, the type `Callable` should be more detailed by giving its input and output types.

File: deepmd/pt/model/model/ener_model.py [hunk 2]
```diff
@@ -95,3 +95,40 @@ def forward_lower(
...
+                "must": False,
+                "high_prec": False,
+            },
+            "atom_pref": {
+                "ndof": 1,
+                "atomic": True,
+                "must": False,
+                "high_prec": False,
+                "repeat": 3,
+            },
```
Comment (njzjz) [inline deepmd/pt/model/model/ener_model.py:132]: It seems to me all of these data are only required for energy loss, but not the energy model (if one doesn't train it). It is unclear to me if it is a good way to put these requirements in the energy model.

File: deepmd/pt/train/training.py [hunk 3]
```diff
@@ -812,28 +854,22 @@ def get_data(self, is_train=True, task_key="Default"):
...
             "coord",
             "atype",
             "box",
-        ]:
-            if item in batch_data:
-                input_dict[item] = batch_data[item]
-            else:
-                input_dict[item] = None
+            "spin",
+        ]
```
Comment (njzjz) [inline deepmd/pt/train/training.py:864]: `fparam` and `aparam` should be added.

File: deepmd/dpmodel/descriptor/make_base_descriptor.py (hunk 1 above)
Comment (iProzd) [reply deepmd/dpmodel/descriptor/make_base_descriptor.py:95]: > What is the purpose of this change?

To pack data for statistics is a very slow and memory-block process, especially when multitask training. Here I used one lazy function to do this, because only inside descriptor&fitting, one can finally decide whether to do data stat or not. If do stat, one can call this lazy function and pack the data for statistics for just once in whole process.

Comment (iProzd) [reply deepmd/dpmodel/descriptor/make_base_descriptor.py]: > Anyway, the type `Callable` should be more detailed by giving its input and output types.

Updated.

File: deepmd/pt/model/model/ener_model.py (hunk 2 above)
Comment (iProzd) [reply deepmd/pt/model/model/ener_model.py]: I've made DataRequirementItem on https://github.com/deepmodeling/deepmd-kit/pull/3359/commits/5db7883457df7d7cdc7523a7354d6fb4a5432cfb. And I'm trying to put these requirements into loss definitions.

**Comment by iProzd** [comment]
https://github.com/iProzd/deepmd-kit/pull/31 fix neighbor-stat in multi-task, but I didn't find out how to set reviewer in this branch.

File: deepmd/dpmodel/descriptor/make_base_descriptor.py [hunk 4]
```diff
@@ -84,8 +86,15 @@ def mixed_types(self) -> bool:
             """
             pass
 
+        @abstractmethod
+        def share_params(self, base_class, shared_level, resume=False):
+            """Share the parameters of self to the base_class with shared_level."""
```
Comment (njzjz) [inline deepmd/dpmodel/descriptor/make_base_descriptor.py:91]: What is resuming? (Should document)

File: deepmd/pt/model/task/dipole.py [hunk 5]
```diff
@@ -89,7 +93,7 @@ def __init__(
         self.r_differentiable = r_differentiable
         self.c_differentiable = c_differentiable
         super().__init__(
-            var_name=var_name,
+            var_name="dipole" if "var_name" not in kwargs else kwargs.pop("var_name"),
```
Comment (njzjz) [inline deepmd/pt/model/task/dipole.py:96]: kwargs.pop("var_name", "dipole")

File: deepmd/pt/model/task/ener.py [hunk 6]
```diff
@@ -138,18 +140,27 @@ def serialize(self) -> dict:
...
+        stat_file_path: Optional[DPPath] = None,
+    ):
         if stat_file_path is not None:
             stat_file_path = stat_file_path / "bias_atom_e"
         if stat_file_path is not None and stat_file_path.is_file():
             bias_atom_e = stat_file_path.load_numpy()
         else:
+            if callable(merged):
+                # only get data for once
+                sampled = merged()
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/ener.py:155]: why only fetch data once ?

File: deepmd/pt/model/model/ener_model.py (hunk 2 above)
Comment (iProzd) [reply deepmd/pt/model/model/ener_model.py]: Fix in https://github.com/deepmodeling/deepmd-kit/pull/3359/commits/fdca653a42c9cc0e2e8490d4dfe476416590f80c

File: deepmd/pt/loss/ener.py [hunk 7]
```diff
@@ -153,3 +160,46 @@ def forward(self, model_pred, label, natoms, learning_rate, mae=False):
         if not self.inference:
             more_loss["rmse"] = torch.sqrt(loss.detach())
         return loss, more_loss
+
+    @property
+    def label_requirement(self) -> List[DataRequirementItem]:
```
Comment (njzjz) [inline deepmd/pt/loss/ener.py:165]: I think it can be dynamic. For example, if no virial loss is set, the virial requirement don't need to be added.

File: deepmd/pt/train/training.py [hunk 8]
```diff
@@ -159,22 +155,43 @@ def get_data_loader(_training_data, _validation_data, _training_params):
...
-                drop_last=False,
-                pin_memory=True,
-            )
-
-            with torch.device("cpu"):
-                validation_data_buffered = BufferedIterator(iter(validation_dataloader))
-            if _training_params.get("validation_data", None) is not None:
-                valid_numb_batch = _training_params["validation_data"].get(
-                    "numb_btch", 1
+            if _validation_data is not None:
```
Comment (wanghan-iapcm) [inline deepmd/pt/train/training.py:158]: the same operations applied to both train and validation. consider writting a function for them rather than copying code.

File: deepmd/pt/utils/dataloader.py [hunk 9]
```diff
@@ -248,9 +256,26 @@ def get_weighted_sampler(training_data, prob_style, sys_prob=False):
...
-    log.info("Generated weighted sampler with prob array: " + str(probs))
+    log.debug("Generated weighted sampler with prob array: " + str(probs))
     # training_data.total_batch is the size of one epoch, you can increase it to avoid too many  rebuilding of iteraters
     len_sampler = training_data.total_batch * max(env.NUM_WORKERS, 1)
     with torch.device("cpu"):
         sampler = WeightedRandomSampler(probs, len_sampler, replacement=True)
     return sampler
+
+
+class LazyFunction:
```
Comment (njzjz) [inline deepmd/pt/utils/dataloader.py:267]: Isn't it just `functools.lru_cache`?

File: deepmd/pt/loss/loss.py [hunk 10]
```diff
@@ -10,3 +18,8 @@ def __init__(self, **kwargs):
     def forward(self, model_pred, label, natoms, learning_rate):
         """Return loss ."""
         raise NotImplementedError
+
+    @property
+    def label_requirement(self) -> List[DataRequirementItem]:
```
Comment (wanghan-iapcm) [inline deepmd/pt/loss/loss.py:23]: better using abstract property
```python
    @property
    @abstractmethod
    def label_requirement(self):
```

File: deepmd/pt/utils/finetune.py [hunk 11]
```diff
@@ -21,7 +21,8 @@ def change_finetune_model_params(
     """
     if multi_task:
         # TODO
-        log.error("finetune mode need modification for multitask mode!")
+        pass
+        # log.error("finetune mode need modification for multitask mode!")
```
Comment (wanghan-iapcm) [inline deepmd/pt/utils/finetune.py:25]: the if statement can be removed.

File: deepmd/pt/model/task/polarizability.py [hunk 12]
```diff
@@ -112,7 +115,7 @@ def __init__(
         ).view(ntypes, 1)
         self.shift_diag = shift_diag
         super().__init__(
-            var_name=var_name,
+            var_name="polar" if "var_name" not in kwargs else kwargs.pop("var_name"),
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/polarizability.py:118]:
```suggestion
            var_name=kwargs.pop("var_name", "polar")
```

File: deepmd/pt/model/task/dipole.py (hunk 5 above)
Comment (wanghan-iapcm) [inline deepmd/pt/model/task/dipole.py:96]:
```suggestion
            var_name=kwargs.pop("var_name", "dipole"),
```

File: deepmd/pt/model/descriptor/se_r.py [hunk 13]
```diff
@@ -151,12 +153,52 @@ def mixed_types(self) -> bool:
         """
         return False
 
-    def compute_input_stats(self, merged: List[dict], path: Optional[DPPath] = None):
+    def share_params(self, base_class, shared_level, resume=False):
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/se_r.py:156]: doc str needed.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/dpmodel/descriptor/make_base_descriptor.py (hunk 1 above)
Comment (iProzd) [reply deepmd/dpmodel/descriptor/make_base_descriptor.py]: fyi @wanghan-iapcm

File: deepmd/pt/model/model/dipole_model.py [hunk 14]
```diff
@@ -92,21 +97,23 @@ def forward_lower(
...
-                "ndof": 3,
-                "atomic": True,
-                "must": False,
-                "high_prec": False,
-                "type_sel": self.get_sel_type(),
-            },
-        }
+    def data_requirement(self) -> List[DataRequirementItem]:
+        data_requirement = [
+            DataRequirementItem(
```
[the remark this answers is not in the document]
Comment (iProzd) [reply deepmd/pt/model/model/dipole_model.py]: It has been moved into loss func.

File: deepmd/pt/train/training.py [hunk 15]
```diff
@@ -119,62 +128,54 @@ def get_opt_param(params):
...
+                (
+                    validation_dataloader,
+                    validation_data_buffered,
+                ) = get_dataloader_and_buffer(_validation_data, _training_params)
+                if _training_params.get("validation_data", None) is not None:
+                    valid_numb_batch = _training_params["validation_data"].get(
+                        "numb_btch", 1
+                    )
+                else:
+                    valid_numb_batch = 1
```
Comment (wanghan-iapcm) [inline deepmd/pt/train/training.py:175]:
```suggestion
                valid_numb_batch = _training_params["validation_data"].get("numb_btch", 1)
```

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

**Review by wanghan-iapcm — APPROVED** [review]

**Review by njzjz — APPROVED** [review]

**Review by CaRoLZhangxy — APPROVED** [review]
